src/sema/types.rs

```diff
@@ -197,37 +204,117 @@ fn type_decl(
-                    def.loc,
-                    format!("struct '{}' has infinite size", def.name),
-                    field.loc,
-                    format!("recursive field '{}'", field.name_as_str()),
-                ));
+/// A struct field is recursive, if it is connected to a cyclic path.
+///
+/// This function checks all structs in the `ns` for any paths leading into the given `node`.
+/// `node` is supposed to be inside a cycle.
+/// All affected struct fields will be flagged as recursive (and infinite size as well, if they are).
+fn check_recursive_struct_field(node: usize, graph: &Graph, ns: &mut Namespace) {
+    for n in 0..ns.structs.len() {
+        for path in all_simple_paths::<Vec<_>, &Graph>(graph, n.into(), node.into(), 0, None) {
+            for (a, b) in path.windows(2).map(|a_b| (a_b[0], a_b[1])) {
+                for edge in graph.edges_connecting(a, b) {
+                    ns.structs[a.index()].fields[*edge.weight()].recursive = true;
+                    if ns.structs[b.index()].fields.iter().any(|f| f.infinite_size) {
+                        ns.structs[a.index()].fields[*edge.weight()].infinite_size = true;
+                    }
+                }
+            }
+        }
+    }
+}
 
-                ns.structs[struct_no].fields[field_no].recursive = true;
-            } else {
-                structs_visited.push(field_struct_no);
-                find_struct_recursion(field_struct_no, structs_visited, ns);
-                structs_visited.pop();
+/// Find all other structs a given user struct may reach.
+///
+/// `edges` is a set with tuples of 3 dimensions. First two are the connecting nodes (struct numbers).
+/// The last dimension is the field number of the first struct where the connection originates.
+fn collect_struct_edges(no: usize, edges: &mut HashSet<(usize, usize, usize)>, ns: &Namespace) {
+    for (field_no, field) in ns.structs[no].fields.iter().enumerate() {
+        for reaching in field.ty.user_struct_no(ns) {
+            if edges.insert((no, reaching, field_no)) {
+                collect_struct_edges(reaching, edges, ns)
             }
         }
     }
 }
 
+/// Checks for
+///   - Structs containing recursive (cycling) fields
+///   - Cycling struct fields of infinite size
+///
+/// The algorithm consists of these steps:
+///   1. All structs in the namespace are parsed into a graph.
+///      Nodes in the graph represent the structs.
+///      Edges develop when a struct encapsulates another struct.
+///      Edges have the originating struct field number as their weight.
+///      So we known from which struct field the connection originated later on.
+///   2. Find all Strongly Connected Components (SCC) in the graph.
+///   3. For any node inside in any SCC, if there is a path from the node to itself, we've detected a cycle.
+///   4. For every cycle, check if it is of infinite size and flag involved struct fields accordingly.
```

Comment:
What do you understand by a cycle of infinite size? Is A->B->C->A an infinite sized cycle? I guess you meant the memory size of the structs forming the cycle, right? The wording here looks like you are talking about the size of the cycle (the quantity of vertices you have in it).

Reply:
Yes of course I meant memory size. I updated the comment so it says memory size. However, completely alien to me how anyone could confuse it like this. It makes no sense. It's impossible to write a contract with structs having an infinite "quantity of vertices" first hand.